File: python/Day05.py
from utils.Stopwatch import runWithStopwatch

text_input = open("../input/day05.txt", "r").readlines()


def part1():
    commands, stacks = parse()
    execute_commands(stacks, commands, crateMover9000)
    print(get_top_crates(stacks))


def part2():
    commands, stacks = parse()
    execute_commands(stacks, commands, crateMover9001)
    print(get_top_crates(stacks))


def parse():
    cutoff_index = text_input.index("\n")
    stacks_input = text_input[:cutoff_index]
    stacks = parseStacks(cutoff_index, stacks_input)
    commands = text_input[cutoff_index + 1:]
    return commands, stacks


def parseStacks(cutoff_index, stacks_input):
    parsing_indexes = range(1, len(stacks_input[0]), 4)
    stacks = []
    for line_index in range(cutoff_index - 1, 0 - 1, -1):
        if line_index == cutoff_index - 1:
            for _ in parsing_indexes:
                stacks.append([])
            continue
        i = 0
        for index in parsing_indexes:
            line = stacks_input[line_index]
            crate = line[index]
            if crate != " ":
                stacks[i].append(crate)
            i += 1
    return stacks


def execute_commands(stacks, commands, execute_func):
    for i in range(0, len(commands)):
        command = commands[i].strip()
        start_index = command.index("move ") + len("move ")
        end_index = command.index(" from ")
        move = command[start_index:end_index]
        start_index = end_index + len(" from ")
        end_index = command.index(" to ")
        from_ = command[start_index:end_index]
        start_index = end_index + len(" to ")
        to = command[start_index:]
        execute_func(stacks, int(from_), int(to), int(move))


def crateMover9000(stacks, from_stack, to_stack, number_of_moving):
    for _ in range(0, number_of_moving):
        moving_crate = stacks[from_stack - 1].pop()
        stacks[to_stack - 1].append(moving_crate)
    return stacks


def crateMover9001(stacks, from_stack, to_stack, number_of_moving):
    moving_crates = []
    for _ in range(0, number_of_moving):
        moving_crates.append(stacks[from_stack - 1].pop())
    for i in range(len(moving_crates) - 1, 0 - 1, -1):
        stacks[to_stack - 1].append(moving_crates[i])
    return stacks


def get_top_crates(stacks):
    top_crates = ""
    for stack in stacks:
        top_crates += stack[len(stack) - 1]
    return top_crates


if __name__ == '__main__':
    runWithStopwatch(part1)
    runWithStopwatch(part2)

File: kotlin/src/main/kotlin/adventofcode2022/Day05.kt
package adventofcode2022

import adventofcode2022.util.runWithStopwatch
import java.io.File

private val input = File("./input/day05.txt").readLines()


private fun part1() {
    val (stacks, commands) = parseInput(input)
    executeCommands(stacks, commands, crateMover9000)
    println(getTopCrates(stacks))
}

private fun part2() {
    val (stacks, commands) = parseInput(input)
    executeCommands(stacks, commands, crateMover9001)
    println(getTopCrates(stacks))
}

fun getTopCrates(stacks: List<ArrayList<Char>>): String {
    var topCrates = ""
    for (i in stacks.indices) {
        topCrates += stacks[i][stacks[i].lastIndex]
    }
    return topCrates
}

fun parseInput(input: List<String>): Pair<List<ArrayList<Char>>, List<String>> {
    val cutoffIndex = input.indexOf("")
    return parseStacks(input.subList(0, cutoffIndex - 1)) to input.subList(cutoffIndex + 1, input.size)
}

fun parseStacks(subList: List<String>): List<ArrayList<Char>> {
    val stacks = List(9) { ArrayList<Char>() }
    for (line in subList.reversed()) {
        for ((stackIndex, elementIndex) in (1 until line.length step 4).withIndex()) {
            val element = line[elementIndex]
            if (element != ' ') {
                stacks[stackIndex].add(element)
            }
        }
    }
    return stacks
}

fun executeCommands(stacks: List<ArrayList<Char>>, commands: List<String>, mover: (stacks: List<ArrayList<Char>>, move: Int, from: Int, to: Int) -> Unit) {
    for (command in commands) {
        var startIndex = command.indexOf("move ") + "move ".length
        var endIndex = command.indexOf(" from ")
        val move = command.substring(startIndex, endIndex).toInt()
        startIndex = endIndex + " from ".length
        endIndex = command.indexOf(" to ")
        val from = command.substring(startIndex, endIndex).toInt()
        startIndex = endIndex + " to ".length
        val to = command.substring(startIndex).toInt()
        mover(stacks, move, from, to)
    }
}

val crateMover9000 = fun (stacks: List<ArrayList<Char>>, move: Int, from: Int, to: Int) {
    val fromIndex = from - 1
    val toIndex = to - 1
    for (i in 0 until move) {
        val crate = stacks[fromIndex][stacks[fromIndex].lastIndex]
        stacks[toIndex].add(crate)
        stacks[fromIndex].removeAt(stacks[fromIndex].lastIndex)
    }
}

val crateMover9001 = fun (stacks: List<ArrayList<Char>>, move: Int, from: Int, to: Int) {
    val fromIndex = from - 1
    val toIndex = to - 1
    val movingCrates = arrayListOf<Char>()
    for (i in 0 until move) {
        movingCrates.add(stacks[fromIndex][stacks[fromIndex].lastIndex])
        stacks[fromIndex].removeAt(stacks[fromIndex].lastIndex)
    }
    for (crate in movingCrates.reversed()) {
        stacks[toIndex].add(crate)
    }
}

fun main() {
    runWithStopwatch { part1() }
    runWithStopwatch { part2() }
}